2022/Q3/artifacts/PoS-tla/Staking.tla

```diff
@@ -78,32 +101,49 @@ MAX_UINT == 2^(256 - 60) - 1
 INITIAL_SUPPLY == 10^(9+18)
 
 \* the number of tokens the validator has staked
+\* @type: () => Int;
 INIT_VALIDATOR_STAKE == 1000000000000000000000
 
 \* the amount of voting power per token
-VOTES_PER_TOKEN == 1
+SLASH_RATE == 1
+
+(*
+* Computes the Max of two numbers.
+*)
+Max(x, y) == IF x > y THEN x ELSE y
+
+(*
+* Computes the Min of two numbers.
+*)
+Min(x, y) == IF x < y THEN x ELSE y
 
 \* Initialize the balances
 Init ==
     /\ balanceOf = [ a \in UserAddrs |-> 
-        IF a \in ValidatorAddrs
-        THEN INITIAL_SUPPLY - INIT_VALIDATOR_STAKE
-        ELSE INITIAL_SUPPLY
-       ]
-    /\ unbonded = [ n \in 1..UnbondingLength, a \in UserAddrs |-> 0 ]
-    /\ delegated = [ n \in 1..UnbondingLength, a \in UserAddrs, b \in ValidatorAddrs |->
-        IF a \in ValidatorAddrs /\ a = b
-        THEN INIT_VALIDATOR_STAKE
-        ELSE 0
-       ]
-    /\ bonded = [ n \in 1..UnbondingLength, a \in ValidatorAddrs |-> INIT_VALIDATOR_STAKE ]
+                     IF a \in ValidatorAddrs
+                     THEN INITIAL_SUPPLY - INIT_VALIDATOR_STAKE
+                     ELSE INITIAL_SUPPLY
+                   ]
+    /\ unbonded = [ a \in UserAddrs, b \in ValidatorAddrs |-> {} ]
+    /\ bonded = [ a \in UserAddrs, b \in ValidatorAddrs |->
+                  IF a = b
+                  THEN { [ id |-> 0, epoch |-> 1, amount |-> INIT_VALIDATOR_STAKE ] }
+                  ELSE {}
+                ]
+    /\ totalDeltas = [ n \in 0..2*UnbondingLength, a \in ValidatorAddrs |-> INIT_VALIDATOR_STAKE ]
+    /\ slashes = [ b \in ValidatorAddrs |-> {} ]
+    /\ enqueuedSlashes = [ n \in -1..UnbondingLength, a \in ValidatorAddrs |-> {}] 
+    /\ frozenValidators = [ n \in 0..UnbondingLength |-> {} ]
+    /\ posAccount = Cardinality(ValidatorAddrs) * INIT_VALIDATOR_STAKE
     /\ nextTxId = 0
-    /\ epoch = 1
+    /\ epoch = UnbondingLength + 1
     /\ txCounter = 0
+    /\ idBonds = 1
+    /\ idUnbonds = 1
+    /\ idSlashes = 1
     /\ lastTx = [ id |-> 0, tag |-> "None", fail |-> FALSE ]
     /\ failed = FALSE
 
-
 \* delegate `amount` tokens to a validator
 Delegate(sender, validator, amount) ==
```

Comment:
If "sender" can be a validator, why do we need "Init" to be so complex? The state Init is trying to set up can be reached with a sequence of "Delegate", can't it?

Reply:
can you be more specific? which part of Init seems complex to you? Are you referring to the fact that validators have a number of tokens bonded from the beginning?

Reply:
Exactly. That can be just the first transition... so the "code" of the model would be simpler.
I know we are resource-limited, but it'd be one transition per validator only, and it would greatly simplify this part

Reply:
The way it is we can have "deeper" traces as the first transition could already be unbond. Nevertheless, I do not really have a strong opinion about it.